### PR TITLE
Implement fixed size transform interpolation buffer.

### PR DESCRIPTION
**Pull request by MichaelGrupp**
While using the TransformInterpolationBuffer in private code I noticed
that it seems to be incomplete. Neither does it have a buffer size
limit, nor does it "remove the oldest transform if the buffer size limit
is exceeded" as stated in the header.

The changes don't break compatibility with existing code:
unless explicitly set, the buffer size is unlimited.

### CODE REVIEW

File: cartographer/transform/transform_interpolation_buffer.h [hunk 1]
```diff
@@ -17,7 +17,8 @@
 #ifndef CARTOGRAPHER_TRANSFORM_TRANSFORM_INTERPOLATION_BUFFER_H_
 #define CARTOGRAPHER_TRANSFORM_TRANSFORM_INTERPOLATION_BUFFER_H_
 
-#include <vector>
+#include <deque>
+#include <numeric>
```
Comment (wohe) [inline cartographer/transform/transform_interpolation_buffer.h:21]: Typo? `#include <limits>` contains `std::numeric_limits`.

File: cartographer/transform/transform_interpolation_buffer.h [hunk 2]
```diff
@@ -27,18 +28,29 @@
...
+constexpr size_t kUnlimitedBufferSize = std::numeric_limits<size_t>::max();
+
 // A time-ordered buffer of transforms that supports interpolated lookups.
+// Unless explicitly set, the buffer size is unlimited.
 class TransformInterpolationBuffer {
  public:
   TransformInterpolationBuffer() = default;
   explicit TransformInterpolationBuffer(
       const mapping::proto::Trajectory& trajectory);
+  explicit TransformInterpolationBuffer(const size_t buffer_size_limit);
```
Comment (wohe) [inline cartographer/transform/transform_interpolation_buffer.h:40]: Not sure this constructor provides much value. Convenience is not necessarily bad, but here it seems to also harm readability as you would call the constructor with just an integer. Calling `SetSizeLimit()` directly after construction makes it much clearer what is happening. Remove?

Comment (MichaelGrupp) [reply cartographer/transform/transform_interpolation_buffer.h:40]: Sounds good.

**Comment by MichaelGrupp** [comment]
I do have a use case for that in an internal package that uses libcartographer. Besides that, it could be also useful for others.

File: cartographer/transform/transform_interpolation_buffer.h [hunk 3]
```diff
@@ -27,18 +28,28 @@
...
+// Unless explicitly set, the buffer size is unlimited.
 class TransformInterpolationBuffer {
  public:
   TransformInterpolationBuffer() = default;
   explicit TransformInterpolationBuffer(
       const mapping::proto::Trajectory& trajectory);
 
+  // Sets the transform buffer size limit and removes old transforms
+  // if it is exceeded.
+  void SetSizeLimit(const size_t buffer_size_limit);
```
Comment (wohe) [inline cartographer/transform/transform_interpolation_buffer.h:43]: Nit: the `const` should be removed as it has no meaning here. It is only useful at the implementation.

Comment (wohe) [reply cartographer/transform/transform_interpolation_buffer.h]: done

**Pull request merged**